### PR TITLE
Dynamic Retrieval of Microsoft.WindowsAppSDK Version for C# Single-Project Templates 

### PR DESCRIPTION
**Pull request by ShashankNay**
**Problem**:

The current approach involves including the WinAppSDK version directly within our templates. This has resulted in a cumbersome process of template updates and submissions for every new WinAppSDK version release. The delay between the release of WinAppSDK updates and the availability of updated templates has, on occasion, exceeded one month, hampering our goal of enhancing release cadence.

**Solution**:

To address this challenge and enhance our release agility, a dynamic method for acquiring the WinAppSDK version has been introduced. Instead of embedding the version directly into templates, the WinAppSDK version is now obtained dynamically through service calls.

**Technical Details:**

The [Visual Studio IWizard interface](https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.templatewizard.iwizard?view=visualstudiosdk-2022) is leveraged to enable the execution of custom code when [users create projects from templates](https://learn.microsoft.com/en-us/visualstudio/extensibility/how-to-use-wizards-with-project-templates?view=vs-2022). This approach facilitates API calls to NuGet for retrieving the latest WinAppSDK version before project creation is completed.

Subscription to the [OnSolutionRestoreFinished](https://learn.microsoft.com/en-us/nuget/visual-studio-extensibility/nuget-api-in-visual-studio#ivsnugetprojectupdateevents-interface) event ensures the timing of our NuGet API calls aligns with the project creation process. This timing is critical due to the asynchronous nature of project loading in Visual Studio, [as noted in the documentation](https://learn.microsoft.com/en-us/nuget/visual-studio-extensibility/nuget-api-in-visual-studio#understanding-the-net-project-systems:~:text=Even%20after%20all,not%20be%20loaded.%22.), which can cause certain NuGet operations to fail if invoked too early in the project or solution loading process.

Through this dynamic version retrieval strategy, the need for altering templates for each WinAppSDK version update is eliminated.

**Future Plans:**

This currently is only supported in the C# scenario. Future plans include using similar techniques for C++ projects.

### CODE REVIEW

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+using Microsoft.VisualStudio.TemplateWizard;
+using Microsoft.VisualStudio.Threading;
+using NuGet.VisualStudio;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+
+namespace WindowsAppSDK.Cs.Extension
```
Comment (Scottj1s) [inline dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs:16]: I'd go ahead and change NS now - maybe something like "WindowsAppSDK.TemplateUtilities" ?

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+{
+
+    public class NuGetPackageInstaller : IWizard
```
Comment (Scottj1s) [inline dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs:19]: nit: lots of blank lines ...

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            _project = project;
+        }
+
+        public void BeforeOpeningFile(ProjectItem projectItem)
+        {
+        }
+
+        public void ProjectItemFinishedGenerating(ProjectItem projectItem)
+        {
+        }
```
Comment (Scottj1s) [inline dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs:69]: use '_' for formal params here too

File: dev/VSIX/ProjectTemplates/Desktop/CSharp/SingleProjectPackagedApp/WinUI.Desktop.Cs.SingleProjectPackagedApp.vstemplate [hunk 4]
```diff
@@ -1,59 +1,68 @@
...
+        <Assembly>Microsoft.VisualStudio.WinRT.TemplateWizards, Version=16.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a</Assembly>
+        <FullClassName>Microsoft.VisualStudio.WinRT.TemplateWizards.UpdatePublisherInManifestWizard</FullClassName>
+    </WizardExtension>
+    <WizardExtension>
+        <Assembly>WindowsAppSDK.Cs.Extension.Dev17, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null</Assembly>
+        <FullClassName>WindowsAppSDK.Cs.Extension.NuGetPackageInstaller</FullClassName>
+    </WizardExtension>
+    <WizardData>
+        <packages>
+            <package id="Microsoft.WindowsAppSDK" />
```
Comment (Scottj1s) [inline dev/VSIX/ProjectTemplates/Desktop/CSharp/SingleProjectPackagedApp/WinUI.Desktop.Cs.SingleProjectPackagedApp.vstemplate:65]: you were able to implement - nice!

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                if (!string.IsNullOrEmpty(packageId))
+                {
+                    _packageId = packageId;
+                }
+            }
+
+            var componentModel = (IComponentModel)Microsoft.VisualStudio.Shell.ServiceProvider.GlobalProvider.GetService(typeof(SComponentModel));
+            IVsPackageInstaller installer = componentModel.GetService<IVsPackageInstaller>();
+            _packageInstaller = installer;
+
```
Comment (Scottj1s) [inline dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs:53]: move to InstallNuGetPackageAsync and remove _packageInstaller

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        private Task InstallNuGetPackageAsync()
+        {
+            try
+            {
+                _packageInstaller.InstallPackage(null, _project, _packageId, "", false);
+            }
+            catch (Exception ex)
+            {
+
+            }
```
Comment (Scottj1s) [inline dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs:93]: are you expecting any exceptions here?  should we plumb thru to VS diagnostics for visibility?

File: dev/VSIX/ProjectTemplates/Desktop/CSharp/SingleProjectPackagedApp/ProjectTemplate.csproj [hunk 7]
```diff
@@ -23,7 +23,6 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.WindowsAppSDK" Version="$WindowsAppSDKNupkgVersion$" />
     <PackageReference Include="Microsoft.Windows.SDK.BuildTools" Version="$WindowsSDKBuildToolsNupkgVersion$" />
     <Manifest Include="$(ApplicationManifest)" />
```
Comment (Scottj1s) [inline dev/VSIX/ProjectTemplates/Desktop/CSharp/SingleProjectPackagedApp/ProjectTemplate.csproj:27]: this should be next to go in future PR

**Comment by Scottj1s** [comment]
I know we want to wrap this up asap, but without doing all of dev/VSIX/ProjectTemplates/Desktop/CSharp/, we can't remove the WinAppSDK freight from the C# project template component VSIX.

**Comment by Scottj1s** [comment]
Great PR description!

File: dev/VSIX/Extension/Cs/Dev17/Standalone/source.extension.vsixmanifest [hunk 8]
```diff
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bpulliam) [inline dev/VSIX/Extension/Cs/Dev17/Standalone/source.extension.vsixmanifest:1]: A diff on the first line indicates a change to line endings or the BOM. Check your editor settings.

File: dev/VSIX/Extension/Cs/Dev17/WindowsAppSDK.Cs.Extension.Dev17.csproj [hunk 9]
```diff
@@ -1,4 +1,4 @@
-﻿<Project ToolsVersion="Current">
+<Project ToolsVersion="Current">
```
Comment (bpulliam) [inline dev/VSIX/Extension/Cs/Dev17/WindowsAppSDK.Cs.Extension.Dev17.csproj:1]: A diff on the first line indicates a change to line endings or the BOM. Check your editor settings.

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            var joinableTaskFactory = new JoinableTaskFactory(ThreadHelper.JoinableTaskContext);
+            joinableTaskFactory.RunAsync(InstallNuGetPackageAsync);
+
+        }
+
+        private Task InstallNuGetPackageAsync()
+        {
+            try
+            {
+                _packageInstaller.InstallPackage(null, _project, _packageId, "", false);
```
Comment (bpulliam) [inline dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs:88]: check for null _packageId value before calling?

File: dev/VSIX/ProjectTemplates/Desktop/CSharp/PackagedApp/WinUI.Desktop.Cs.PackagedApp.vstemplate [hunk 11]
```diff
@@ -1,37 +1,37 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bpulliam) [inline dev/VSIX/ProjectTemplates/Desktop/CSharp/PackagedApp/WinUI.Desktop.Cs.PackagedApp.vstemplate:1]: Check editor settings

**Review by bpulliam — COMMENTED** [review]
One other overall comment - check the versions of dependencies, I see 16.0, 17,3, 17.4, and 17.5. Do we take this opportunity to sync those?

File: dev/VSIX/Extension/Cs/Dev17/WindowsAppSDK.Cs.Extension.Dev17.csproj [hunk 12]
```diff
@@ -39,11 +39,10 @@
     </None>
```
Comment (bpulliam) [inline dev/VSIX/Extension/Cs/Dev17/WindowsAppSDK.Cs.Extension.Dev17.csproj:39]: line 31 above should read 'Dev17 templates should target .NET 6' (not 5)

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs (hunk 10 above)
Comment (ShashankNay) [reply dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs]: Thinking of putting a default of `Microsoft.WindowsAppSDK` if for some reason there's a problem retrieving this value via the WizardData

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs (hunk 6 above)
Comment (ShashankNay) [reply dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs:86]: Should we go with some sort of console-based error message, or a popup of sorts that lets the user know that WASDK failed to install and that they should install the latest version via a link?

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs (hunk 10 above)
Comment (ShashankNay) [reply dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs]: Changed the logic to check for this an display an error

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs (hunk 6 above)
Comment (ShashankNay) [reply dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs:107]: I've done both here: a message displayed to the user with instructions on how to install the latest WASDK should they be unable to, along with a log error.

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+			_nugetProjectUpdateEvents.SolutionRestoreFinished -= OnSolutionRestoreFinished;
+			var joinableTaskFactory = new JoinableTaskFactory(ThreadHelper.JoinableTaskContext);
+			joinableTaskFactory.RunAsync(InstallNuGetPackageAsync);
+
+		}
+		private Task InstallNuGetPackageAsync()
+		{
+            if (string.IsNullOrEmpty(_packageId))
+            {
+                string message = "Failed to install the NuGet package. The package ID provided in the template configuration is either missing or invalid. Please ensure the template is correctly configured with a valid package ID.";
```
Comment (Scottj1s) [inline dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs:80]: generally, I prefer to see diagnostics nearer the point of discovery (line 51 in this case)

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+                DisplayMessageToUser(message, "Error", OLEMSGICON.OLEMSGICON_CRITICAL);
```
Comment (Scottj1s) [inline dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs:81]: maybe "Configuration Error" to be more specific (as you have "Installation Error" below)

File: dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+            try
+            {
+                installer.InstallPackage(null, _project, _packageId, "", false);
+            }
+            catch (Exception ex)
+            {
+                string errorMessage = "Failed to install the Microsoft.WindowsAppSDK package. You can try installing it manually from: https://www.nuget.org/packages/Microsoft.WindowsAppSDK";
+                DisplayMessageToUser(errorMessage, "Installation Error", OLEMSGICON.OLEMSGICON_CRITICAL);
+
+                string logMessage = $"Failed to install Microsoft.WindowsAppSDK package. Exception details: \n" +
```
Comment (Scottj1s) [inline dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs:95]: for future reference, see String.Format

**Review by Scottj1s — CHANGES_REQUESTED** [review]
other than the BOM change, my last few comments are nits - take or leave

**Review by Scottj1s — APPROVED** [review]

File: dev/VSIX/ProjectTemplates/Desktop/CSharp/PackagedApp/WinUI.Desktop.Cs.PackagedApp.vstemplate (hunk 11 above)
Comment (bpulliam) [reply dev/VSIX/ProjectTemplates/Desktop/CSharp/PackagedApp/WinUI.Desktop.Cs.PackagedApp.vstemplate:1]: There are no actual changes (outside of BOM and line/file endings) to this file, so it should not be in the PR.

File: dev/VSIX/ProjectTemplates/Desktop/CSharp/PackagedApp/WapProj/WinUI.Desktop.Cs.WapProj.vstemplate [hunk 16]
```diff
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (bpulliam) [inline dev/VSIX/ProjectTemplates/Desktop/CSharp/PackagedApp/WapProj/WinUI.Desktop.Cs.WapProj.vstemplate:1]: There are no actual changes (outside of BOM and line/file endings) to this file, so it should not be in the PR.

**Review by bpulliam — APPROVED** [review]

**Comment by Scottj1s** [comment]
:shipit:

**Pull request merged**